### ISSUE TITLE
ChannelClosed exception when daemon is running for a long time

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida_core#1366]
Potentially related to #1335 

I was running the daemon with twelve workers on theossrv5 and after 12 hours of heavy workchain submissions, a workchain crashed with the following stack trace:

```
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base_process.py", line 569, in _run_task
    return fn(*args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base_process.py", line 304, in _run
    self._action_command(command)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base_process.py", line 316, in _action_command
    self.transition_to(ProcessState.RUNNING, command.continue_fn, *command.args)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/base/state_machine.py", line 148, in transition_to
    self.state_machine.transition_to(state, *args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/processes.py", line 446, in transition_to
    self.on_entered(initial_state)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/plumpy/plumpy/processes.py", line 393, in on_entered
    subject="state_changed.{}.{}".format(from_label, self.state.value)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/communicator.py", line 329, in broadcast_send
    return self._message_publisher.broadcast_send(body, sender, subject, correlation_id)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/communicator.py", line 50, in broadcast_send
    self.action_message(message)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 303, in action_message
    message.send(self)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 111, in send
    publisher.publish_msg(self.message, defaults.BROADCAST_TOPIC)
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 326, in publish_msg
    mandatory=mandatory
  File "/home/aiida/.virtualenvs/aiida_sdb/src/kiwipy/kiwipy/rmq/messages.py", line 342, in do_publish
    publish_channel.publish(*args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 2127, in publish
    self._flush_output(self._message_confirmation_result.is_ready)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1250, in _flush_output
    *waiters)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 455, in _flush_output
    self._impl.ioloop.poll()
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 245, in poll
    self._poller.poll()
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 935, in poll
    self._dispatch_fd_events(fd_event_map)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 625, in _dispatch_fd_events
    handler(fileno, events)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 395, in _handle_events
    self._handle_read()
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 449, in _handle_read
    self._on_data_available(data)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/connection.py", line 1938, in _on_data_available
    self._process_frame(frame_value)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/connection.py", line 2059, in _process_frame
    if self._process_callbacks(frame_value):
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/connection.py", line 2040, in _process_callbacks
    frame_value)  # Args
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/callback.py", line 60, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/callback.py", line 92, in wrapper
    return function(*args, **kwargs)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "/home/aiida/.virtualenvs/aiida_sdb/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1316, in _on_channel_closed
    method.reply_text)
ChannelClosed: (404, "NOT_FOUND - no exchange 'aiida-69b2078be6d74abd97446c7371983e62.messages' in vhost '/'")
```

**Comment by greschd** [comment]
I don't think time is the relevant factor here -- I managed to trigger the same exception almost instantaneously, by running a very heavy calculation (uses a few GB of input data). Traceback:

```
04/03/2018 02:38:40 PM <5541> pika.adapters.base_connection: [ERROR] Error event 25, None
04/03/2018 02:38:40 PM <5541> pika.adapters.base_connection: [CRITICAL] Tried to handle an error where no error existed
04/03/2018 02:38:40 PM <5541> pika.callback: [ERROR] Calling <bound method BlockingChannel._on_channel_closed of <BlockingChannel impl=<Channel number=1 CLOSED conn=<SelectConnection OPEN socket=('127.0.0.1', 35110)->('127.0.0.1', 5672) params=<URLParameters host=127.0.0.1 port=5672 virtual_host=/ ssl=False>>>>> for "1:Channel.Close" failed
Traceback (most recent call last):
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1316, in _on_channel_closed
    method.reply_text)
ChannelClosed: (404, "NOT_FOUND - no exchange 'aiida-d26c044f49e14a1ea8e3f4d23d1170cb.messages' in vhost '/'")
04/03/2018 02:38:40 PM <5541> aiida.aiida_wannier90.calculations.Wannier90Calculation: [REPORT] [788|JobProcess_aiida_wannier90.calculations.Wannier90Calculation|on_except]: Traceback (most recent call last):
  File "/media/greschd/large_data/projects/Cd3As2/tbextraction/source/aiida_core/aiida/work/job_processes.py", line 242, in action_command
    self.scheduler_update()
  File "/media/greschd/large_data/projects/Cd3As2/tbextraction/source/aiida_core/aiida/work/job_processes.py", line 289, in scheduler_update
    data=UPDATE_SCHEDULER_COMMAND)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/plumpy/plumpy/base/state_machine.py", line 148, in transition_to
    self.state_machine.transition_to(state, *args, **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/plumpy/plumpy/processes.py", line 458, in transition_to
    self.on_entered(initial_state)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/plumpy/plumpy/processes.py", line 402, in on_entered
    subject='state_changed.{}.{}'.format(from_label, self.state.value)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/communicator.py", line 329, in broadcast_send
    return self._message_publisher.broadcast_send(body, sender, subject, correlation_id)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/communicator.py", line 50, in broadcast_send
    self.action_message(message)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 303, in action_message
    message.send(self)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 111, in send
    publisher.publish_msg(self.message, defaults.BROADCAST_TOPIC)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 326, in publish_msg
    mandatory=mandatory
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/messages.py", line 342, in do_publish
    publish_channel.publish(*args, **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 2127, in publish
    self._flush_output(self._message_confirmation_result.is_ready)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1250, in _flush_output
    *waiters)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 455, in _flush_output
    self._impl.ioloop.poll()
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 245, in poll
    self._poller.poll()
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 935, in poll
    self._dispatch_fd_events(fd_event_map)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 625, in _dispatch_fd_events
    handler(fileno, events)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 395, in _handle_events
    self._handle_read()
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 449, in _handle_read
    self._on_data_available(data)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/connection.py", line 1938, in _on_data_available
    self._process_frame(frame_value)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/connection.py", line 2059, in _process_frame
    if self._process_callbacks(frame_value):
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/connection.py", line 2040, in _process_callbacks
    frame_value)  # Args
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 60, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 92, in wrapper
    return function(*args, **kwargs)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/callback.py", line 236, in process
    callback(*args, **keywords)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1316, in _on_channel_closed
    method.reply_text)
ChannelClosed: (404, "NOT_FOUND - no exchange 'aiida-d26c044f49e14a1ea8e3f4d23d1170cb.messages' in vhost '/'")
```

**Comment by greschd** [comment]
Somewhere down the line I also get 

```
04/03/2018 02:38:42 PM <5541> tornado.application: [ERROR] Future exception was never retrieved: Traceback (most recent call last):
  File "/home/greschd/.virtualenvs/cd3as2_aiida/local/lib/python2.7/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/tasks.py", line 140, in _on_task
    self._task_finished(method.delivery_tag, props.correlation_id, props.reply_to, response)
  File "/home/greschd/.virtualenvs/cd3as2_aiida/src/kiwipy/kiwipy/rmq/tasks.py", line 157, in _task_finished
    self._channel.basic_ack(delivery_tag=delivery_tag)
AttributeError: 'NoneType' object has no attribute 'basic_ack'
```
but I suspect that's just a consequence of the first error.

**Comment by sphuber** [comment]
The second exception is indeed just a manifestation of the first problem, when the connection is closed the channel is set to `None`. I think this fix still addresses a valid problem, we tested this extensively over the weekend, but I just think you have found another failure mode that can cause the connection to be shutdown. I am guessing that since you mention heavy input data if it is being serialized and sent to RabbitMQ that may cause it to fail. Or it is the serialization by `yaml` that topples.

**Comment by greschd** [comment]
Here's part of the RabbitMQ log:

```
=ERROR REPORT==== 3-Apr-2018::14:36:39 ===
closing AMQP connection <0.13438.0> (127.0.0.1:34048 -> 127.0.0.1:5672):
missed heartbeats from client, timeout: 60s

=INFO REPORT==== 3-Apr-2018::14:38:40 ===
accepting AMQP connection <0.14427.0> (127.0.0.1:35110 -> 127.0.0.1:5672)

=INFO REPORT==== 3-Apr-2018::14:38:40 ===
connection <0.14427.0> (127.0.0.1:35110 -> 127.0.0.1:5672): user 'guest' authenticated and granted access to vhost '/'

=ERROR REPORT==== 3-Apr-2018::14:38:40 ===
Channel error on connection <0.14427.0> (127.0.0.1:35110 -> 127.0.0.1:5672, vhost: '/', user: 'guest'), channel 1:
operation basic.publish caused a channel exception not_found: no exchange 'aiida-d26c044f49e14a1ea8e3f4d23d1170cb.messages' in vhost '/'

=INFO REPORT==== 3-Apr-2018::14:38:41 ===
accepting AMQP connection <0.14440.0> (127.0.0.1:35112 -> 127.0.0.1:5672)
```

I wonder if the large input data is what triggers the timeout / heartbeat problem, and that then somehow triggers this error?

**Comment by sphuber** [comment]
That may very well be the case, if the heartbeat is missed (I think twice) RabbitMQ assumes the subscriber is dead and closes the channel. If the transfer is taking longer than the timeout this may very be the problem. I notice that the timeout is still 60 seconds. The fix we implemented in upped this value to 600 seconds. We also changed some other channel/queue policies which will require a reset of the RabbitMQ server. Can you do the following and try your heavy input example again:

```
sudo rabbitmqctl stop_app
sudo rabbitmqctl reset
sudo rabbitmqctl start_app
verdi daemon restart --reset
```

**Comment by greschd** [comment]
Just to make it clear: The job is successfully launched and runs _before_ the exception comes along and just puts the process in the excepted state.

**Comment by greschd** [comment]
I'm still using ``devel run_daemon`` #1317. Is there an equivalent to ``restart --reset`` I have to run, or is just starting a new daemon enough?

**Comment by greschd** [comment]
Ah, I guess I should pull the fixes first :sweat_smile:

**Comment by greschd** [comment]
Seems the reason is that the minimum between the server and client values is picked, see https://github.com/pika/pika/pull/666 (right at the end). I'll just bump the value in my server configuration for now, but I guess we probably want a fix for that _before_ pika 1.0 comes out.

**Comment by greschd** [comment]
Ok, seems solved now after a bit of a struggle getting rabbitmq to accept my config. Here's what I learned:
- The config file format changed from RabbitMQ 3.6 to 3.7. Since I still have 3.6, I needed to use the old format (file attached, remove ``.txt`` ending).
- For actually loading the config file, I needed to run ``sudo /etc/init.d/rabbitmq-server start``, ``sudo /etc/init.d/rabbitmq-server stop``. When just running ``stop_app``, ``reset``, ``start_app``, it just complained that it didn't find any (there was a mistake in the config file, so that might be why).

[rabbitmq.config.txt](https://github.com/aiidateam/aiida_core/files/1873021/rabbitmq.config.txt)

**Comment by greschd** [comment]
So in the case I had, there were four calculations to be submitted. I suspect that just a single submission wouldn't have triggered the timeout. Is there a way that we can keep the heartbeat alive between submitting calculations? I can see a batch of submissions taking longer than 10 minutes on a bad connection.

**Comment by sphuber** [comment]
Wait, why did you have to change the config of RabbitMQ? I never touched any configuration

**Comment by greschd** [comment]
Because it was set to 60 seconds server-side, and pika picks the _lower_ of the two values when negotiating the heartbeat (see the issue I linked above). Maybe there is no default set on your server installation?

**Comment by greschd** [comment]
You can easily check what the _actual_ heartbeat is by going to the connection from the management GUI.

**Comment by greschd** [comment]
FWIW, there's a beta version for pika 1.0 already, so I guess it will be solved from that side in the not-too-distant future.

**Comment by sphuber** [comment]
That is interesting to know. I have the default version `3.5.7` on Ubuntu 16.04.4 LTS. Apparently there the behavior is different as I did not have to change anything in the server settings. If this has been changed in newer versions of RabbitMQ we should probably find a way to override this from aiida-core. We don't want users to have to mess with the configuration if we can at all avoid it. Thanks for the info :+1:

**Comment by greschd** [comment]
When pika 1.0 is out, I can test again if that solves the problem without manual config on my machine. I don't know if there's a good way to override the behaviour from within AiiDA.

It's curious that you're not seeing this behaviour -- from what I read, the timeout default was set to 60 seconds in version ``3.5.5``. Maybe they differ in how it's negotiated between client and server, though.

**Comment by sphuber** [comment]
Tentatively fixed in #1370
